docs: Add Maintaining AUR packages with Renovate user story (#21413)
renovatebot/renovate: docs/usage/user-stories/maintaining-aur-packages-with-renovate.md

<!-- hide table of contents in navigation sidebar -->
<style>
.md-nav--primary .md-nav__link[for=__toc] ~ .md-nav {
    display: none;
}
</style>

# Maintaining AUR packages with Renovate

> This article was written by [Jamie Magee](https://github.com/JamieMagee) and originally published on [Jamie Magee's blog](https://jamiemagee.co.uk/blog/maintaining-aur-packages-with-renovate/).

<!-- prettier-ignore -->
!!! note
    Jamie Magee helps to maintain Renovate.
    They obviously like Renovate, and want you to use it.

> One big advantage that Arch Linux has over other distributions, apart from being able to say “BTW I use arch”, is the Arch User Repository (AUR).

Author: ```suggestion
One big advantage that Arch Linux has over other distributions, apart from being able to say “BTW I use Arch.”, is the Arch User Repository (AUR).
```

Arch is a proper noun, so we should capitalize it. And sentences should end with a full stop.

It’s a community-driven repository with over 80,000 packages.
If you’re looking for a package, chances are you'll find it in the AUR.

Keeping all those packages up to date, takes a lot of manual effort by a lot of volunteers.
People have created and used tools, like [`urlwatch`](https://github.com/thp/urlwatch) and [`aurpublish`](https://github.com/eli-schwartz/aurpublish), to let them know when upstream releases are cut and automate some parts of the process.
I know I do.
But I wanted to automate the entire process.
I think [Renovate](https://github.com/renovatebot/renovate/) can help here.

## Updating versions with Renovate

Renovate is an automated dependency update tool.
You might have seen it opening pull requests on GitHub and making updates for npm or other package managers, but it’s a lot more powerful than just that.

Renovate has a couple of concepts that I need to explain first: [datasources](../modules/datasource/index.md) and [managers](../modules/manager/index.md).
Datasources define where to look for new versions of a dependency.

> Renovate comes with over 50 different datasources, but the one that is important for AUR packages is the [`git-tags` datasource](https://docs.renovatebot.com/modules/datasource/#git-tags-datasource).

Author: The source article says "over 40 different datasources", but I'm counting 50+ datasources on the datasources homepage in the docs. Please check my counting. 😄

Reviewer: I started writing this article a while ago. That probably accounts for the difference.

Managers are the Renovate concept for package managers.
There isn’t an AUR or `PKGBUILD` manager, but there is a [regex manager](https://docs.renovatebot.com/modules/manager/regex/) that I can use.

I can create a `renovate.json` configuration with the following regex manager configuration:

```json
{
  "regexManagers": [
    {
      "fileMatch": ["(^|/)PKGBUILD$"],
      "matchStrings": [
        "pkgver=(?<currentValue>.*) # renovate: datasource=(?<datasource>.*) depName=(?<depName>.*)"
      ],
      "extractVersionTemplate": "^v?(?<version>.*)$"
    }
  ]
}
```

Breaking that down:

- The `fileMatch` setting tells Renovate to look for any `PKGBUILD` files in a repository
- The `matchStrings` is the regex format to extract the version, datasource, and dependency name from the `PKGBUILD`
- The `extractVersionTemplate` is to handle a “v” in front of any version number that is sometimes added to Git tags

And here’s an extract from the PKGBUILD for the [bicep-bin](https://aur.archlinux.org/packages/bicep-bin) AUR package that I maintain:

```bash
pkgver=0.15.31 # renovate: datasource=github-tags depName=Azure/bicep
```

> Here I’m configuring Renovate to use the [`github-tags`](https://docs.renovatebot.com/modules/datasource/github-tags/) datasource and to look in the [Azure/bicep GitHub repository](https://github.com/Azure/bicep) for new versions.

Author: ```suggestion
Here I’m configuring Renovate to use the [`github-tags`](https://docs.renovatebot.com/modules/datasource/github-tags/) datasource and to look in the [`Azure/bicep` GitHub repository](https://github.com/Azure/bicep) for new versions.
```

Let's monospace the repository name.

> That means it’ll look in the [list of tags in that repository](https://github.com/Azure/bicep/tags) for any new versions.

Author: ```suggestion
That means it’ll look in the [list of tags for the `Azure/bicep` repository](https://github.com/Azure/bicep/tags) for any new versions.
```

Use a better link name.

If Renovate finds any new versions, it’ll automatically update the `PKGBUILD` and open a pull request with the updated version.

So I’ve automated the `PKGBUILD` update, but that’s only half of the work.
The checksums and `.SRCINFO` must be updated before pushing to the AUR.

> Unfortunately, Renovate can’t do that ([yet](https://github.com/renovatebot/renovate/tree/feat/arch-linux-manager)), but GitHub Actions can!

Author: @JamieMagee Is there an issue we can link to? It feels a bit odd to link to a work in progress branch.

Reviewer: Yep! https://github.com/renovatebot/renovate/issues/16923

Author: ```suggestion
Unfortunately, Renovate can’t do that (yet, see [Renovate issue #16923](https://github.com/renovatebot/renovate/issues/16923)), but GitHub Actions can!
```

You can [check out the full code on GitHub](https://github.com/JamieMagee/aur-packages/tree/main/.github/actions/aur).

```yaml
- name: Validate package
  if: ${{ env.pkgbuild != '' }}
  uses: ./.github/actions/aur
  with:
    action: 'updpkgsums'
    pkgname: ${{ env.pkgbuild }}
```

Finally, once the `PKGBUILD` and `.SRCINFO` are updated I need to commit that change back to the pull request.

```yaml
- name: Commit
  if: ${{ env.pkgbuild != '' }}
  uses: stefanzweifel/git-auto-commit-action@3ea6ae190baf489ba007f7c92608f33ce20ef04a # v4.16.0
  with:
    file_pattern: '*/PKGBUILD */.SRCINFO'
```

Check out [this pull request for `bicep-bin`](https://github.com/JamieMagee/aur-packages/pull/62) where Renovate opened a pull request, and my GitHub Actions workflow updated the `b2sums` in the `PKGBUILD` and updated the `.SRCINFO`.

But why stop there?
Let’s talk about publishing.

## Publishing to the AUR

Each AUR package is its own Git repository.
So to update a package in the AUR, I only need to push a new commit with the updated `PKGBUILD` and `.SRCINFO`.

> Thankfully, [KSXGitHub](https://github.com/KSXGitHub) created the [github-actions-deploy-aur](https://github.com/KSXGitHub/github-actions-deploy-aur) GitHub Action to streamline the whole process.

Author: ```suggestion
Thankfully, [KSXGitHub](https://github.com/KSXGitHub) created the [`github-actions-deploy-aur` GitHub Action](https://github.com/KSXGitHub/github-actions-deploy-aur) to streamline the whole process.
```

Monospace for the repository name, and include the GitHub Action part in the link name.

If I create a new GitHub Actions workflow to publish to the AUR, I can reuse the first two steps from my previous workflow to check out the repository and find the updated package.

> Then all I need to do is to use the [github-actions-deploy-aur](https://github.com/KSXGitHub/github-actions-deploy-aur) GitHub Action:

Author: ```suggestion
Then all I need to do is to use the `github-actions-deploy-aur` GitHub Action:
```

We already link to the action on line 151, so we can probably drop the second link.

```yaml
- name: Publish package
  uses: KSXGitHub/github-actions-deploy-aur@065b6056b25bdd43830d5a3f01899d0ff7169819 # v2.6.0
  if: ${{ env.pkgbuild != '' }}
  with:
    pkgname: ${{ env.pkgbuild }}
    pkgbuild: ${{ env.pkgbuild }}/PKGBUILD
    commit_username: ${{ secrets.AUR_USERNAME }}
    commit_email: ${{ secrets.AUR_EMAIL }}
    ssh_private_key: ${{ secrets.AUR_SSH_PRIVATE_KEY }}
```

### All together now

> If you own any AUR packages and want to automate some of the maintenance burden, check out my [AUR packages template GitHub repository](https://github.com/JamieMagee/aur-packages-template/).

Author: ```suggestion
If you own any AUR packages and want to automate some of the maintenance burden, check out [my AUR packages template GitHub repository](https://github.com/JamieMagee/aur-packages-template/).
```

Put the word `my` in the link.

It contains all of the steps I showed in this blog post.

> And if you want to see how it works in practice, check out my [AUR packages GitHub repository](https://github.com/JamieMagee/aur-packages).

Author: ```suggestion
And if you want to see how it works in practice, check out [my AUR packages GitHub repository](https://github.com/JamieMagee/aur-packages).
```

Same thing here.
